### PR TITLE
Command to manage profiles

### PR DESCRIPTION
**Pull request by genoud**
sync: retrieve profiles components  form a Salesforce org with all their
configurations, even unchecked user permission and object permissions
for object with no access
reconcile: cleanup profiles components to make them compliant with a
Salesforce org. For example, profiles fetch from production may not
deploy on sandboxes or scratch orgs due to user permissions that does
not exist on those orgs (ManageSandboxes) or object that are not
supported on those orgs (ScratchOrgInfo)
merge: retrieve profiles from a development org (Sandbox or scratch org
and merge their configurations on the profile metadata files, keeping settings
only available on production. The use of the option -m can also help
retrieve only permissions on specific metadata (CustomObject, Layout or
CustomTab)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -765,4 +758,76 @@ EXAMPLE
   }
 ```
 
- 
\ No newline at end of file
+## `sfpowerkit:profile:sync`
+
+retrieve profiles from the salesforce org with all their configurations. Run this command preferably again the production org. Use the merge equivalent again other type of orgs.
```
Comment (azlam-abdulsalam) [inline README.md]: Can you make it clear what this command does?I believe for this command to be run you should have the full set of metadata

File: README.md [hunk 2]
```diff
@@ -765,4 +758,76 @@ EXAMPLE
...
+
+EXAMPLES
+  $ sfdx sfpowerkit:profile:reconcile  --folder force-app
+  $ sfdx sfpowerkit:profile:reconcile  --folder force-app,module2,module3 -u sandbox
+  $ sfdx sfpowerkit:profile:reconcile  -u myscratchorg
+```
+
+## `sfpowerkit:profile:merge`
+
+retrieve profiles from a development environment and merge it with the profile configurations
```
Comment (azlam-abdulsalam) [inline README.md]: Can you provide more details on how this command actually works? what will be stripped of?

File: src/metadatainfo.json [hunk 3]
```diff
@@ -0,0 +1,908 @@
+{
```
Comment (azlam-abdulsalam) [inline src/metadatainfo.json]: Can you move this file from the src folder to somewhere else? may be resources?

File: src/shared/appUtils.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
+import BaseUtils from "./baseUtils";
```
Comment (azlam-abdulsalam) [inline src/shared/appUtils.ts]: As these files are mainly to support profile fetching and not necessarily shared between any other commands, Lets create a folder to support profile

File: README.md (hunk 1 above)
Comment (genoud) [reply README.md]: No need to have the full set of metadata. The command read the metadata using the `metadata.readSync` function. Then complete userPermissions for permissions not retrieved (Those unchecked in the org). The list of supported permissions is computed using a Describe on the PermissionSet Sobject. Then the command complete ObjectPermission for objects with No Read Access. The list of supported objects is computed by running the query `SELECT SobjectType, count(Id) From ObjectPermissions Group By sObjectType`
For standard profiles, the List of UserPermissions is removed as salesforce does not allow to edit user permissions for standard Profiles. So no need to retrieve them.

File: README.md (hunk 2 above)
Comment (genoud) [reply README.md]: This command do the same thing as the sync command but instead overwriting  the profile metadata files, it update it configuration with those retrieve from the org. This avoid loosing profiles settings from one org to another.

Running the sync from the production org and the merge from a scratch org will result in our profile metadata containing settings that exist in both orgs but also settings that are available only in each org. 

The reconcile is then necessary to clean up profile to make them compatible to deploy in any org by removing what is not supported in the org.

**Comment by genoud** [comment]
> We are not able to get this command to work on a folder structure having src/core it was throwing undefined

This error is not due to the folder structure, but to the fact that the folder src/core does not contain any profile.

File: src/shared/appUtils.ts (hunk 4 above)
Comment (genoud) [reply src/shared/appUtils.ts]: Yes, those files mainly support profiles fetching now. But will be necessary when implementing the same feature for permission sets.

**Comment by genoud** [comment]
> @genoud.
> 
> Can you rebase based on latest master.

Done

File: README.md [hunk 5]
```diff
@@ -677,32 +677,25 @@ EXAMPLE
...
 
 ```
 USAGE
-  $ sfdx sfpowerkit:source:picklist:generatepatch [-p <string>] [-d <string>] [-f <boolean>]  [-r <boolean>] 
+  $ sfdx sfpowerkit:source:picklist:generatepatch [-p <string>] [-d <string>] [--json] [--loglevel trace|debug|info|warn|error|fatal|TRACE|DEBUG|INFO|WARN|ERROR|FATAL]
 
 OPTIONS
   -d, --objectsdir=objectsdir                                                       Path for Objects folder located in project
-  -p, --package=package                                                             Name of the package to generate the picklist 
-  patch
```
Comment (azlam-abdulsalam) [inline README.md]: The readme is outdated, please update on top of the  new one

File: src/shared/appUtils.ts (hunk 4 above)
Comment (azlam-abdulsalam) [reply src/shared/appUtils.ts]: okie,lets move it to another folder still, shared is becoming too cluttered

**Comment by azlam-abdulsalam** [comment]
> > We are not able to get this command to work on a folder structure having src/core it was throwing undefined
> 
> This error is not due to the folder structure, but to the fact that the folder src/core does not contain any profile.

We tried moving profile to the folder,but still failed, @manisfdcsfdx  please recheck

File: README.md (hunk 5 above)
Comment (genoud) [reply README.md]: Done

**Comment by azlam-abdulsalam** [comment]
I have moved around the files to a different folder for clarity

File: src/shared/appUtils.ts (hunk 4 above)
Comment (azlam-abdulsalam) [reply src/shared/appUtils.ts]: Moved to shared profile utils

File: README.md (hunk 2 above)
Comment (azlam-abdulsalam) [reply README.md]: Okie, understood, lets get the readme little elaborated

File: README.md (hunk 1 above)
Comment (azlam-abdulsalam) [reply README.md]: Lets elaborate the readme to make it bit more clear

**Pull request closed without merging**